tests/arch/arm/arm_thread_swap/src/arm_syscalls.c

```diff
@@ -239,3 +240,64 @@ void test_arm_syscalls(void)
 /**
  * @}
  */
+
+
+void z_impl_test_arm_cpu_write_reg(void)
+{
+	/* User thread CPU write registers system call for testing
+	 *
+	 * Verify the following
+	 * - Write 0xDEADBEEF values during system call into registers
+	 * - In main test we will read that registers to verify
+	 * that all of them were scrubbed and do not contain any sensitive data
+	 */
+
+	/* Part below is made to test that kernel scrubs CPU registers
+	 * after returning from the system call
+	 */
+	__asm__ volatile (
+		"ldr r0, =0xDEADBEEF;\n\t"
+		"ldr r1, =0xDEADBEEF;\n\t"
+		"ldr r2, =0xDEADBEEF;\n\t"
+		"ldr r4, =0xDEADBEEF;\n\t"
+		);
+}
+
+static inline void z_vrfy_test_arm_cpu_write_reg(void)
+{
+	z_impl_test_arm_cpu_write_reg();
+}
+#include <syscalls/test_arm_cpu_write_reg_mrsh.c>
+
+/**
+ * @brief Test CPU scrubs registers after system call
+ *
+ * @details - Call string nlen and right before exit from
+ * the system call function write into registers 0xDEADBEEF value
+ * - Then in main test function below check registers values,
+ * if no 0xDEADBEEF value detected, that means CPU scrubbed registers
+ * before exit from the system call.
+ *
+ * @ingroup kernel_memprotect_tests
+ */
+void test_syscall_cpu_scrubs_regs(void)
+{
+	long arm_reg_val[4];
+
+	test_arm_cpu_write_reg();
+
+	__asm__ volatile (
+		"\t ldr %%r0,=0" : "=r"(arm_reg_val[0]));
+	__asm__ volatile (
+		"\t ldr %%r1,=0" : "=r"(arm_reg_val[1]));
+	__asm__ volatile (
+		"\t ldr %%r2,=0" : "=r"(arm_reg_val[2]));
+	__asm__ volatile (
+		"\t ldr %%r4,=0" : "=r"(arm_reg_val[3]));
```

Comment:
This is not correct. You are setting these regs to 0 here.

If any, it should be:
```
__asm__ volatile("mov %0, rN" : "=r"(arm_reg_val[N]));
```

Reply:
Or use str  @stephanosio , correct?

Reply:
@maksimmasalski  I believe what you need to do here is to store r0-r3 to memory and then check what is going on.